### PR TITLE
fix(transformers): gate _tie_weights_nemo on tie_word_embeddings flag

### LINKED ISSUE #2941: fix(transformers): _tie_weights_nemo force-ties untied models (ignores tie_word_embeddings)
## Summary

`_tie_weights_nemo` (`nemo_automodel/_transformers/model_init.py:994-1007`) re-ties `lm_head.weight` to the input embedding for any model exposing `_nemo_tied_weights_keys` — **without checking** `config.tie_word_embeddings`. For untied models this silently discards or corrupts trained weights.

Introduced in [#1817](<https://github.com/NVIDIA-NeMo/Automodel/issues/1817>) ("fix: tie weights outside `_init_model`") together with the patched `post_init` in `_transformers/utils.py` that *sets* `_nemo_tied_weights_keys`. [#1817](<https://github.com/NVIDIA-NeMo/Automodel/issues/1817>)'s intent was correct — re-tie **tied** pre-v5 remote-code models (Nemotron-Flash-1B) after checkpoint loading breaks the construction-time alias — but the re-tie fires unconditionally. Pre-dates the NVIDIA-NeMo/Automodel#2512/NVIDIA-NeMo/Automodel#2896 guard work; surfaced during the tie-support audit for NVIDIA-NeMo/Automodel#2935.

## Affected paths

Two ways a model gets `_nemo_tied_weights_keys`:

1. **Static class attribute**: `ernie4_5` dense + MoE (`components/models/ernie4_5/model.py:416,520`). An ERNIE config with `tie_word_embeddings=False` branches correctly in `__init__`, then gets force-tied anyway at the end of `_init_model` (`model_init.py:1040-1041`).
2. **Dynamic (broader)**: the patched `post_init` (`_transformers/utils.py`) sets it for **every pre-v5 remote-code model with list-form** `_tied_weights_keys`. In pre-v5 HF convention, declaring the keys does **not** mean the model is tied — HF's own `tie_weights()` checks the config flag. Many models declare the keys but ship untied checkpoints; all of them are force-tied when loaded through `NeMoAuto*`.

## Observable failure modes

* **Fresh** `from_pretrained` **of an untied model** (silent): HF loads the trained `lm_head.weight`, then `_tie_weights_nemo` aliases it away to `embed_tokens.weight`. No error, no spike — logits simply diverge from the HF reference from the first forward; the model is effectively tied although the checkpoint (and the class it will be served with) is untied.
* **Resume/load of a checkpoint with distinct** `lm_head`**/embedding weights** (loud): after construction the two params share one storage; `load_state_dict`/DCP copies both checkpoint tensors into the same storage — last writer wins, corrupting embeddings and/or head → **loss spike on resume**.

## Proposed fix

Gate the re-tie on the controlling config flag (reusing the canonical resolver from `components/checkpoint/utils.py`):

```python
def _tie_weights_nemo(model):
    if not hasattr(model, "_nemo_tied_weights_keys"):
        return
    if not get_controlling_tie_word_embeddings(model.config, type(model).__name__):
        return  # untied config: the loaded separate lm_head is authoritative
    ...
```

Tied models (Nemotron-Flash) keep the [#1817](<https://github.com/NVIDIA-NeMo/Automodel/issues/1817>) behavior unchanged. Add regression tests for both manifestations: (a) untied list-form remote-code model keeps distinct storage + exact HF logits parity through the `NeMoAuto*` path; (b) save-then-reload of distinct weights round-trips losslessly (no spike); (c) tied model still re-ties (no [#1817](<https://github.com/NVIDIA-NeMo/Automodel/issues/1817>) regression).

## Relationship to NVIDIA-NeMo/Automodel#2935

Independent and complementary: NVIDIA-NeMo/Automodel#2935's constructor guards cover registered classes only. The dynamic path here affects **unregistered** HF/remote-code classes, which no per-class declaration can reach — this gate is the only line of defense for them. (The ERNIE static case additionally becomes unreachable once NVIDIA-NeMo/Automodel#2935 demotes ERNIE to tied-only.)

### PR DESCRIPTION
**Pull request by yuhezhang-ai**
# What does this PR do ?

Fixes #2941: `_tie_weights_nemo` re-tied `lm_head.weight` to the input embeddings for any model exposing
`_nemo_tied_weights_keys`, without consulting the controlling `config.tie_word_embeddings` flag.
`_nemo_tied_weights_keys` records candidate tied keys from pre-v5/list-form `_tied_weights_keys`; it does not prove
that the checkpoint is actually tied.

Impact is real for shipped untied remote-code force-HF recipes. Confirmed affected:

- `examples/llm_finetune/nemotron/nemotron_nano_4b_squad.yaml`
- `examples/llm_finetune/nemotron/nemotron_nano_9b_squad.yaml`
- `examples/llm_finetune/nemotron/nemotron_nano_9b_squad_peft.yaml`

Those recipes load Nemotron Nano 4B/9B through `trust_remote_code: true` + `force_hf: true`. The hub configs set
`tie_word_embeddings=False`, the checkpoints contain separate trained `lm_head.weight` and
`backbone.embeddings.weight`, and the recipe compatibility patch converts remote list-form
`_tied_weights_keys = ["lm_head.weight"]` into `_nemo_tied_weights_keys`. Before this PR, `_tie_weights_nemo` then
forced an alias and discarded the trained head.

The fix gates the re-tie on the canonical `get_controlling_tie_word_embeddings` resolver. Tied models, such as the
Nemotron-Flash case #1817 was written for, keep the existing re-tie behavior unchanged. Models without a `config`
attribute conservatively fall back to re-tying.

## Before / after

**Real recipe-path source load: `nvidia/NVIDIA-Nemotron-Nano-9B-v2`.**

Using the same recipe condition that exposes the bug (`apply_cache_compatibility_patches()` + `trust_remote_code=True`
+ `force_hf=True`), compare a raw HF source load against a NeMoAuto source load on a seeded 128-token prompt:

| branch | `lm_head` aliased? | max abs diff vs HF | mean abs diff vs HF | cosine |
|---|---:|---:|---:|---:|
| main | `true` | `22.5625` | `2.9893` | `-0.0224` |
| this PR | `false` | `0.25` | `0.0141` | about `1.000` |

This is the silent manifestation: the model loads successfully, but logits diverge from the raw HF reference immediately
on main because the trained untied `lm_head` is aliased away.

**Unit/mechanism coverage.**

The regression tests fail on main and pass on this branch:

- `test_untied_config_keeps_separate_lm_head`
- `test_untied_state_dict_roundtrip_is_lossless`
- `test_tied_config_reties`
- `test_missing_config_still_reties`

Targeted result on this branch: `4 passed`.

**Resume smoke note.**

A real 9B FSDP2 DCP recipe resume smoke was also run. This PR produced a checkpoint at step 3, then main and this PR
both resumed one step from that same checkpoint. Both resumed step 4 at `loss=0.466808`.

That result means the DCP checkpoint resume path was correct in this smoke. It does not cover the bug fixed here: on
main, the model is already wrong immediately after the initial NeMoAuto source load, before any training checkpoint is
saved or resumed. This is also why the current checkpoint robustness test can miss the issue: it captures reference
logits from the already-loaded NeMo model, then validates checkpoint reloads against that reference. We should add a
separate source-load parity check that compares raw HF source load vs NeMoAuto source load before training starts.

# Changelog

- `_tie_weights_nemo` now returns early when the controlling `tie_word_embeddings` flag is `False`, preserving
  separately loaded `lm_head.weight` for untied models.
- Added `TestTieWeightsNemoConfigGate` regression tests: untied preservation, tied re-tie, missing-config fallback, and
  lossless construct-then-load round-trip.

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed the [Contributor
      Guide](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md).
- [x] Did you write any new necessary tests?
  - `tests/unit_tests/_transformers/test_model_init.py::TestTieWeightsNemoConfigGate`
- [x] Did you add or update any necessary documentation?
  - No public API change.
- [x] Does this PR introduce a breaking change?
  - No. This restores the checkpoint-config contract for untied models and preserves tied behavior.

# Additional Information
* Related to #2935 but intentionally kept as a minimal correctness fix instead of refactoring the full tie-weight
  pipeline.
* Introduced by the #1817-era workaround for pre-v5/list-form `_tied_weights_keys`. That workaround is still needed for
  tied remote-code models; this PR only adds the missing config gate for untied checkpoints.
* Existing checkpoint robustness coverage does not exercise this exact source-load failure. It captures reference logits
  from the already-constructed NeMo model, then compares checkpoint reloads against that reference. If the source load is
  already corrupted, the corrupted model can become the baseline. We should add an opt-in raw-HF-source vs
  NeMoAuto-source parity check for remote-code/force-HF recipes.

### CODE REVIEW

**Comment by yuhezhang-ai** [comment]
/ok to test [9199d1e](https://github.com/NVIDIA-NeMo/Automodel/pull/2942/commits/9199d1e54294845ad3a95f943b463b2886e94aca)

**Review by akoumpa — APPROVED** [review]
